x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -202,43 +227,83 @@ public Automaton allowedActionsMatcher(String index) {
+                if (indexCheck) {
+                    boolean actionCheck = group.checkAction(action);
+                    granted = granted || actionCheck;
+                    // mapping updates are allowed for certain privileges on indices and aliases (but not on data streams),
+                    // outside of the privilege definition
+                    boolean bwcMappingActionCheck = isMappingUpdateAction && false == isDataStream && false == isBackingIndex &&
+                            containsPrivilegeThatGrantsMappingUpdatesForBwc(group);
+                    bwcGrantMappingUpdate = bwcGrantMappingUpdate || bwcMappingActionCheck;
+                    if (actionCheck || bwcMappingActionCheck) {
+                        // propagate DLS and FLS permissions over the concrete indices
+                        for (String index : concreteIndices) {
+                            Set<FieldPermissions> fieldPermissions = fieldPermissionsByIndex.computeIfAbsent(index, (k) -> new HashSet<>());
+                            fieldPermissionsByIndex.put(indexOrAlias, fieldPermissions);
+                            fieldPermissions.add(group.getFieldPermissions());
+                            DocumentLevelPermissions permissions =
+                                    roleQueriesByIndex.computeIfAbsent(index, (k) -> new DocumentLevelPermissions());
+                            roleQueriesByIndex.putIfAbsent(indexOrAlias, permissions);
+                            if (group.hasQuery()) {
+                                permissions.addAll(group.getQuery());
+                            } else {
+                                // if more than one permission matches for a concrete index here and if
+                                // a single permission doesn't have a role query then DLS will not be
+                                // applied even when other permissions do have a role query
+                                permissions.setAllowAll(true);
+                            }
+                        }
+                        if (false == actionCheck) {
+                            for (String privilegeName : group.privilege.name()) {
+                                if (PRIVILEGE_NAME_SET_BWC_ALLOW_MAPPING_UPDATE.contains(privilegeName)) {
+                                    bwcDeprecationLogActions.add(() -> {
+                                        deprecationLogger.deprecate("[" + indexOrAlias + "] mapping update for ingest privilege [" +
+                                                privilegeName + "]", "the mapping update action [" + action + "] on the [" +
+                                                indexOrAlias + "] index, is granted by the [" + privilegeName + "] privilege," +
+                                                " but the privilege has been tightened to not allow it in the next major release");
+                                    });
```

Comment:
How about

> the index privilege [] allowed the update mapping action [] on index [], this privilege will not permit mapping updates in the next major release - users who require access to update mappings must be granted explicit privileges

If you don't like it we can merge as-is, & come back and reword it later.

Reply:
This is clearer, I used idioms such as "tightening privileges".